File: app/frontend/modules/survey/models/Preview/Results/AudioResponse.ts
import _ from 'lodash'

type Answers = Array<{ value: string, media_id: number, filename: string }>

interface Result {
  meta: object
  answers: Answers
}

export default class AudioResponse {
  result: Result

  constructor (result: Result) {
    this.result = result
  }

  meta = (data: object): void => {
    this.result.meta = data
  }

  answer = (url: string, mediaId: number, filename: string): void => {
    if (url) {
      this.result.answers = [{ value: url, media_id: mediaId, filename }]
    } else {
      this.result.answers = []
    }
  }

  results = (): Answers => this.result.answers

  requiredValidation = (): number => _.compact(_.map(this.result.answers, 'value')).length
}

File: lib/imports/assessments/questions/audio_response.rb
# frozen_string_literal: true

module Imports
  module Assessments
    module Questions
      class AudioResponse
        def self.build_answers(data, question, _use_scoring = false, assign)
          return nil if data.compact.blank?

          media_record = MediaResponse.find_by_encoded_id(data[1])
          media_record_to_import = MediaResponses::FindOrCreateMediaResponse.call!(media_record, assign, question)

          return unless media_record_to_import

          {
            answers: [{
              value: media_record_to_import.asset.url,
              file_name: media_record_to_import.filename,
              media_id: media_record_to_import.id
            }],
            question_id: question.id
          }
        end
      end
    end
  end
end
